libeos-updater-util/avahi-service-file.c

```diff
@@ -38,65 +41,131 @@ const gchar * const eos_avahi_v1_ostree_path = "eos_ostree_path";
 const gchar * const eos_avahi_v1_head_commit_timestamp = "eos_head_commit_timestamp";
 
 static gchar *
-txt_records_to_string (const gchar **txt_records)
+txt_records_to_string (GVariant *txt_records)
 {
+  GVariantIter iter;
   g_autoptr(GString) str = NULL;
-  gsize idx;
+  GVariant *txt_value;
+  const gchar *txt_key;
+
+  g_assert (g_variant_type_equal (g_variant_get_type (txt_records),
+                                  G_VARIANT_TYPE ("a(sv)")));
 
   str = g_string_new ("");
 
-  for (idx = 0; txt_records[idx] != NULL; idx++)
+  g_variant_iter_init (&iter, txt_records);
+  while (g_variant_iter_loop (&iter, "(&sv)", &txt_key, &txt_value))
     {
-      g_autofree gchar *record_escaped = g_markup_escape_text (txt_records[idx], -1);
-      g_string_append_printf (str, "    <txt-record>%s</txt-record>\n",
-                              record_escaped);
+      const GVariantType *txt_value_type = g_variant_get_type (txt_value);
+      gboolean txt_value_is_binary = g_variant_type_equal (txt_value_type, G_VARIANT_TYPE_BYTESTRING);
+      gboolean txt_value_is_text = g_variant_type_equal (txt_value_type, G_VARIANT_TYPE_STRING);
+      g_autofree gchar *txt_key_escaped = g_markup_escape_text (txt_key, -1);
+
+      g_assert (txt_value_is_binary || txt_value_is_text);
+
+      if (txt_value_is_binary)
+        {
+          gsize txt_value_data_length;
+          gconstpointer txt_value_data = g_variant_get_fixed_array (txt_value,
+                                                                    &txt_value_data_length,
+                                                                    1);
+          g_autofree gchar *txt_value_encoded = NULL;
+          g_autofree gchar *txt_value_escaped = NULL;
+
+          if (txt_value_data_length + strlen (txt_key) + 1 > 255)
+            g_warning ("The TXT record with key %s is longer than 255 bytes, it will be truncated!",
```

Comment:
Hmm, bloom size is one of the options we pass to the generator function, so adding an assertion here would mean that the users of the `eos_ostree_avahi_service_file_generate` function will need to do the math themselves (this 255 - 3 - strlen(key)). Unless we implement this `rb0`, `rb1`, `rbN` thing.